metricbeat/module/system/process/network/tracker.go

```diff
@@ -0,0 +1,259 @@
+		// as support is missing from go-sysinfo. This means we can't support /hostfs settings
+		procWatcher: watcher,
+
+		log: logp.L(),
+	}
+
+	return tracker, nil
+}
+
+// Update the tracker with the given counts
+func (track *Tracker) Update(packetLen int, proto applayer.Transport, proc *common.ProcessTuple) {
+	track.updateChan <- CounterUpdateEvent{pktLen: packetLen, TransProtocol: proto, Proc: proc}
+}
+
+// Get data for a given PID
+func (track *Tracker) Get(pid int) PacketData {
+	req := RequestCounters{Pid: pid, Resp: make(chan PacketData)}
+	track.reqChan <- req
+	got := <-req.Resp
+	return got
+}
+
+// Stop the tracker
+func (track *Tracker) Stop() {
+	track.stopChan <- struct{}{}
+}
+
+// Track is a non-blocking operation that starts a packet sniffer, and the underlying
+// tracker that correlates packet data with pids
+func (track *Tracker) Track(ctx context.Context) error {
+	helperContext, cancel := context.WithCancel(ctx)
+	go func() {
+		if !track.testmode {
+			err := StartPacketHandle(helperContext, track.procWatcher, track)
+			if err != nil {
+				track.log.Errorf("error starting packet capture: %s", err)
+				ctx.Done()
+			}
+		}
+	}()
+
+	go func() {
+		track.garbageCollect(helperContext)
+	}()
+
+	go func() {
+		for {
+			select {
+			case update := <-track.updateChan:
+				track.updateInternalTracking(update)
+			case req := <-track.reqChan:
+				track.dataMut.RLock()
+				if proc, ok := track.procData[req.Pid]; ok {
+					req.Resp <- proc
+				} else {
+					req.Resp <- PacketData{}
+				}
+				track.dataMut.RUnlock()
+			case <-ctx.Done():
+				cancel()
+				return
+			case <-track.stopChan:
+				cancel()
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// As far as I can see, the only reliable way to do garbage collection
+// is to check each pid on a timer. Simply relying on a timestamp or counter could
+// result in false positives for PIDs that only occasionally create network traffic.
+// This method operates in its own thread, and will check the list of known PIDs on a timer.
+func (track *Tracker) garbageCollect(ctx context.Context) {
+	// the timer should be set to a fairly large number; if we garbage-collect more frequently there's
+	// a risk that we clean up a dead PID before its been reported.
+	ticker := time.NewTicker(track.gctime)
+	for {
+		select {
+		case <-ticker.C:
+			//copy total proc list so we don't hold the mutex for any longer than needed
+			track.dataMut.RLock()
+			keys := make([]int, 0, len(track.procData))
+			for k := range track.procData {
+				keys = append(keys, k)
+			}
+			track.dataMut.RUnlock()
+			keysToDelete := []int{}
+			for _, key := range keys {
+				found, _ := gcPidFetch(ctx, int32(key))
+				if !found {
+					keysToDelete = append(keysToDelete, key)
+				}
+			}
+			if len(keysToDelete) > 0 {
+				track.dataMut.Lock()
+				for _, key := range keysToDelete {
+					delete(track.procData, key)
+				}
```

Comment:
Hmm, yah. In my experience, PID reuse does happen on some containerized workflows. I'm a bit reluctant to carry the lock through that entire operation, since the `gcPidFetch` operation will take some amount of time, and can potentially bottleneck things. The worst case here is that the data for the "reused" PID is off by a few bytes, as it will just get re-added the next time it reports data. Still not a great solution, though.

Other idea might be to add some kind of `updated` tag to the `PacketData` object, to check if the updated time is newer than the "check" time.